devito/profiler.py

```diff
@@ -115,133 +108,251 @@ def add_profiling(self, code, name, byte_size=4, omp_flag=None, to_ignore=None):
+        self.code = code
+        self.name = name
+        self.openmp = openmp
+        self.byte_size = byte_size
+
+        self._var_count = 0
+        self.to_ignore = [
+            "int",
+            "float",
+            "double",
+            "F",
+            "e",
+            "fabsf",
+            "powf",
+            "floor",
+            "ceil",
+            "temp",
+            "i",
+            "t",
+            "p",  # This one shouldn't be here.
+                  # It should be passed in by an Iteration object.
+                  # Added only because tti_example uses it.
+        ] + to_ignore
+        self.seen = set()
+        self.temps = set()
+
+    def run(self):
+        """
+        Calculates the total operation intensity of the code provided.
+        If needed, lets the C code calculate it.
         """
         loads = defaultdict(int)
+        num_flops = 0
+        pragmas = []
 
-        for elem in code:
+        for elem in self.code:
             if isinstance(elem, Assign):
-                assign_flops = self._get_assign_flops(elem, loads, to_ignore)
-                self.oi[name] += assign_flops
-                self.flops_defaults[name] += assign_flops
+                num_flops += self._handle_assign(elem, loads)
             elif isinstance(elem, For):
-                for_flops = self._get_for_flops(
-                    name, elem, loads, to_ignore, temps)
-                self.oi[name] += for_flops
+                num_flops += self._handle_for(elem, loads, pragmas)
+                pragmas = []
             elif isinstance(elem, Block):
-                block_oi, block_flops = self._get_block_oi_and_flops(
-                    name, elem, loads, to_ignore, temps)
-                self.oi[name] += block_oi
-                self.flops_defaults[name] += block_flops
-            elif isinstance(elem, Pragma) and self.openmp:
-                self.add_reduction(elem, name, temps)
+                num_flops += self._handle_block(elem, loads, pragmas)[0]
+                pragmas = []
+            elif isinstance(elem, Pragma):
+                pragmas.append(elem)
             else:
                 # no op
                 pass
 
         load_val_sum = sum(loads.values())
-        self.oi[name] = float(self.oi[name]) / (size*(len(loads) + loads["stores"] - 1))
-        self.oi_low[name] = float(self.oi[name]) / (size*load_val_sum)
-        self.total_load_count[name] = load_val_sum - loads["stores"]
 
-    def _get_for_flops(self, name, loop, loads, to_ignore, temps):
+        oi = float(num_flops) / (self.byte_size*(len(loads) + loads["stores"] - 1))
+        oi_low = oi / (self.byte_size*load_val_sum)
+        num_loads = load_val_sum - loads["stores"]
+
+        return oi, oi_low, num_loads, self.temps
+
+    def _get_loop_reduction(self, variable):
+        return " reduction(+:%s)" % variable
+
+    def _add_reduction(self, pragma, name):
```

Comment:
Can we not just squash this method and ``get_loop_reduction`` into a single method ``_handle_reduction`` ?